exec/src/main/scala/cilib/Runner.scala

```diff
@@ -1,28 +1,118 @@
+                            eval: RVar[NonEmptyList[A] => Objective[A]])
+final case class Progress[A] private (algorithm: String,
+                                      problem: String,
+                                      seed: Long,
+                                      iteration: Int,
+                                      env: Env,
+                                      value: A)
+
 object Runner {
 
-  // Fixme: need to unify these two functions into a single generic one -> the pattern is there
-  def repeat[F[_], A, B](n: Int,
-                         alg: Kleisli[Step[A, ?], F[B], F[B]],
-                         collection: RVar[F[B]]): Step[A, F[B]] =
-    Step
-      .pointR(collection)
+  def repeat[M[_]: Monad, F[_], A](n: Int, alg: Kleisli[M, F[A], F[A]], collection: RVar[F[A]])(
+      implicit M: MonadStep[M]): M[F[A]] =
+    M.pointR(collection)
       .flatMap(coll =>
-        (1 to n).toStream.foldLeftM[Step[A, ?], F[B]](coll) { (a, _) =>
+        (1 to n).toStream.foldLeftM[M, F[A]](coll) { (a, _) =>
           alg.run(a)
       })
 
-  def repeatS[F[_], A, S, B](n: Int,
-                             alg: Kleisli[StepS[A, S, ?], F[B], F[B]],
-                             collection: RVar[F[B]]): StepS[A, S, F[B]] =
-    StepS
-      .pointR(collection)
-      .flatMap(coll =>
-        (1 to n).toStream.foldLeftM[StepS[A, S, ?], F[B]](coll) { (a, _) =>
-          alg.run(a)
-      })
+  def staticProblem[S, A](
+      name: String Refined NonEmpty,
+      eval: RVar[NonEmptyList[A] => Objective[A]],
+      rng: RNG
+  ): Process[Task, Problem[A]] = {
+    val (_, e) = eval.run(rng)
+    Process.constant(Problem(name, Unchanged, RVar.point(e)))
+  }
+
+  def problem[S, A](name: String Refined NonEmpty,
+                    state: RVar[S],
+                    next: S => RVar[(S, Eval[NonEmptyList, A])])(
+      env: Stream[Env],
+      rng: RNG
+  ): Process[Task, Problem[A]] = {
+    def go(s: S, c: Eval[NonEmptyList, A], e: Stream[Env], r: RNG): Process[Task, Problem[A]] =
+      e match {
+        case Stream.Empty => Process.empty
+        case h #:: t =>
+          h match {
+            case Unchanged =>
+              Process.emit(Problem(name, h, c.eval)) ++ go(s, c, t, r)
+
+            case Change =>
+              val (rng2, (s1, c1)) = next(s).run(r)
+              Process.emit(Problem(name, h, c1.eval)) ++ go(s1, c1, t, rng2)
+          }
+      }
+
+    val (rng2, (s2, e)) = state.flatMap(next).run(rng)
+    go(s2, e, env, rng2)
+  }
+
+  /**
+    *  Interpreter for algorithm execution
+    */
+  def foldStep[F[_], A, B](initalConfig: Environment[A],
+                           rng: RNG,
+                           collection: RVar[F[B]],
+                           alg: Algorithm[Kleisli[Step[A, ?], F[B], F[B]]], // Simplify?
+                           env: Process[Task, Problem[A]],
+                           onChange: F[B] => RVar[F[B]]): Process[Task, Progress[F[B]]] = {
```

Comment:
What is `onChange`?

Reply:
It's a function that gets called when the environment changes to adjust the collection. Thanks for seeing this - I missed adding it to the implementation. Will correct it now.